multik-cuda/src/main/kotlin/org/jetbrains/kotlinx/multik/cuda/GpuArray.kt

```diff
@@ -0,0 +1,125 @@
+package org.jetbrains.kotlinx.multik.cuda
+
+import jcuda.CudaException
+import jcuda.Pointer
+import jcuda.runtime.JCuda
+import jcuda.runtime.cudaError.cudaErrorMemoryAllocation
+import jcuda.runtime.cudaMemcpyKind
+import mu.KotlinLogging
+import org.jetbrains.kotlinx.multik.ndarray.data.Dimension
+import org.jetbrains.kotlinx.multik.ndarray.data.MultiArray
+import java.lang.ref.Cleaner
+import java.util.*
+import java.util.concurrent.LinkedBlockingQueue
+
+
+private val logger = KotlinLogging.logger {}
+
+internal class GpuArray private constructor(
+    private val rawData : Any,
+    val hostDataPtr : Pointer,
+    val deviceDataPtr : Pointer,
+    val byteSize : Long
+) {
+    var isLoaded = true
+        private set
+
+    fun copyFromGpu() {
+        checkResult(JCuda.cudaMemcpy(hostDataPtr, deviceDataPtr, byteSize, cudaMemcpyKind.cudaMemcpyDeviceToHost))
+    }
+
+    private fun free() {
+        if (!isLoaded)
+            throw IllegalStateException("Trying to free memory that it is already freed")
+
+        logger.debug { "Freeing GPU memory. Data: $rawData, size: ${byteSizeToString(byteSize)}" }
+        checkResult(JCuda.cudaFree(deviceDataPtr))
+
+        isLoaded = false
+    }
+
+    companion object {
+        private const val CACHE_INITIAL_CAPACITY = 16
+        private const val CACHE_LOAD_FACTOR = 0.75f
+
+        private val cache = Collections.synchronizedMap(
+            LinkedHashMap<Any, GpuArray>(CACHE_INITIAL_CAPACITY, CACHE_LOAD_FACTOR, true)
+        )
+
+        private val deleteQueue = LinkedBlockingQueue<Any>()
+        private val cleaner = Cleaner.create()
+
+        fun assertAllLoaded(vararg arrays: GpuArray) {
+            if (!arrays.all { it.isLoaded })
+                throw OutOfMemoryError("Not all arrays have are loaded in the GPU memory")
```

Comment:
Wrong exception, more suitable here IllegalStateException:
```kotlin
check (arrays.all { it.isLoaded} ) { ... }
```

Reply:
Fixed in newer commits